source/dagger/gameplay/editor/editor_main.cpp

```diff
@@ -0,0 +1,381 @@
+        {
+            Engine::Registry().clear();
+        }
+
+        if (ImGui::MenuItem("Save"))
+        {
+            Engine::Dispatcher().trigger<SaveGameSystem<ECommonSaveArchetype>::SaveRequest>(
+                SaveGameSystem<ECommonSaveArchetype>::SaveRequest{ m_Filename });
+        }
+
+        if (ImGui::MenuItem("Load"))
+        {
+            Engine::Dispatcher().trigger<SaveGameSystem<ECommonSaveArchetype>::LoadRequest>(
+                SaveGameSystem<ECommonSaveArchetype>::LoadRequest{ m_Filename });
+        }
+
+        ImGui::EndMenu();
+    }
+}
+
+void EditorToolSystem::OnKeyboardEvent(KeyboardEvent event_)
+{
+    if (event_.key == EDaggerKeyboard::KeyTab && event_.action == EDaggerInputState::Released)
+        if (Input::IsInputDown(EDaggerKeyboard::KeyLeftControl))
+            m_IsInEditor = !m_IsInEditor;
+}
+
+void EditorToolSystem::Run()
+{
+    if (m_IsInEditor)
+    {
+        auto& knob = m_Registry.get<Sprite>(m_Focus);
+        knob.color.a = 1;
+        auto& focus = m_Registry.get<EditorFocus>(m_Focus);
+
+        if (Input::IsInputDown(EDaggerMouse::MouseButton3))
+        {
+            knob.position = Vector3{ Input::CursorPositionInWorld(), 0 };
+            focus.dirty = true;
+        }
+
+        if (Input::IsInputDown(EDaggerKeyboard::KeyLeftShift))
+        {
+            auto& reg = Engine::Registry();
+            if (reg.valid(m_Selected.entity))
+            {
+                auto& sprite = reg.get<Sprite>(m_Selected.entity);
+                knob.position = Vector3{ Input::CursorPositionInWorld(), 0 };
+                sprite.position = knob.position;
+            }
+        }
+
+        if (!Input::IsInputDown(EDaggerMouse::MouseButton1) && focus.dirty)
+        {
+            m_Targets.clear();
+            focus.dirty = false;
+            Engine::Registry().view<Sprite, SaveGame<ECommonSaveArchetype>>()
+                .each([&](Entity entity_, const Sprite& sprite_, const SaveGame<ECommonSaveArchetype>& save_)
+                    {
+                        const auto left = sprite_.position.x - (sprite_.size.x / 2) * sprite_.scale.x;
+                        const auto top = sprite_.position.y - (sprite_.size.y / 2) * sprite_.scale.y;
+
+                        const auto right = sprite_.position.x + (sprite_.size.x / 2) * sprite_.scale.x;
+                        const auto bottom = sprite_.position.y + (sprite_.size.y / 2) * sprite_.scale.y;
+
+                        if (knob.position.x >= left && knob.position.y >= top &&
+                            knob.position.x <= right && knob.position.y <= bottom)
+                        {
+                            m_Targets.push_back(EditorFocusTarget{ entity_, sprite_.image->Name() });
+                        }
+                    });
+        }
+    }
+    else
+    {
+        auto& knob = m_Registry.get<Sprite>(m_Focus);
+        knob.color.a = 0;
+    }
+}
+
+void EditorToolSystem::OnRenderGUI()
```

Comment:
This function is 200 lines long with a lot of nested conditions and that makes it really difficult to read. Would be nice if it could be broken down into a lot of helper functions with descriptive names.

Reply:
Check if this is more to your liking now :)

Reply:
I like it much better now!